### PR TITLE
Add option for `Pypen` to be serializable

### PR DESCRIPTION
**Pull request by veotani**
Tested with something like this if it makes sense for you:

```py
from creepy.subprocess import Pypen

process = Pypen('main', serializable=True)
initial_state = process.request('get_state')
child_process = Pypen('child')
modified_state = child_process.request('modify_and_get_state', process)
state = process.request('get_state')
assert state == modified_state
assert state != initial_state
```

### CODE REVIEW

File: creepy/subprocess/pypen.py [hunk 1]
```diff
@@ -68,14 +76,26 @@ def __init__(self, args, hash: Optional[str] = None, **kwargs):
...
+            self._send, self._recv = secure_channel(send, recv, self._cipher_name, self._symmetric_key)
         except Exception:
             self.detach()
 
     @property
     def pid(self):
         return self._process.pid
 
+    def __getstate__(self):
+        return self._out_path, self._in_path, self._cipher_name, self._symmetric_key
```
Comment (progbeat) [inline creepy/subprocess/pypen.py:89]: Maybe to raise some error if not serializable?

File: creepy/subprocess/pypen.py [hunk 2]
```diff
@@ -108,6 +131,22 @@ def detach(self):
...
+
+
+def _make_fifo(path: str | None = None):
+    """Open FIFO at specified path or create and open a new one."""
+    if path is None:
+        path = os.path.join('/tmp', secrets.token_urlsafe(16))
+        os.mkfifo(path)
+    in_fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
+    out_fd = os.open(path, os.O_WRONLY)
+    fcntl.fcntl(in_fd, fcntl.F_SETFL, os.O_RDONLY)
```
Comment (progbeat) [inline creepy/subprocess/pypen.py:148]: Please comment such complex to understand lines, explain why do you use specific flags, etc…

File: creepy/subprocess/pypen.py [hunk 3]
```diff
@@ -108,6 +131,25 @@ def detach(self):
...
+def _make_fifo(path: str | None = None):
+    """Open FIFO at specified path or create and open a new one."""
+    if path is None:
+        path = os.path.join('/tmp', secrets.token_urlsafe(16))
+        os.mkfifo(path)
+    # Open the FIFO for reading. `os.O_NONBLOCK` is used in order to prevent blocking the process when opening.
+    in_fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
+    # Since we already opened the FIFO for reading, following open won't block.
+    out_fd = os.open(path, os.O_WRONLY)
+    # Set the file descriptor to blocking mode in order to synchronize the execution of the two processes.
```
Comment (veotani) [inline creepy/subprocess/pypen.py]: ChatGPT doesn't agree with my comment:
> When you use `fcntl.fcntl(in_fd, fcntl.F_SETFL, os.O_RDONLY)`, you're not just removing the `os.O_NONBLOCK` flag; you're actually setting the file descriptor's flags to `os.O_RDONLY` exclusively. This action overrides all existing flags, potentially leading to unintended consequences.

However, there is no such operation as "remove flag" in `fcntl`. It proposes following approach:
```py
# Retrieve the current flags
current_flags = fcntl.fcntl(in_fd, fcntl.F_GETFL)

# Clear the os.O_NONBLOCK flag
new_flags = current_flags & ~os.O_NONBLOCK

# Apply the updated flags
fcntl.fcntl(in_fd, fcntl.F_SETFL, new_flags)
```
As far as I see it is the same logic as I do in my function and it's not hard to understand with new comments. If I'm missing something or a clarification is still needed please let me know.

File: creepy/subprocess/common.py [hunk 4]
```diff
@@ -58,6 +58,10 @@ def secure_alice(send, recv):
     ))
     cipher_name = recv().decode()
     symmetric_key = onetime_private_key.decrypt(recv(), _OAEP_PADDING)
+    return cipher_name, symmetric_key
+
+
+def secure_channel(send, recv, cipher_name, symmetric_key):
```
Comment (romawhite47) [inline creepy/subprocess/common.py:64]: It's not clear how `secure_channel` differs from `_secure_channel`.

**Comment by romawhite47** [comment]
> Tested with something like this if it makes sense for you:
> 
> ```python
> from creepy.subprocess import Pypen
> 
> process = Pypen('main', serializable=True)
> initial_state = process.request('get_state')
> child_process = Pypen('child')
> modified_state = child_process.request('modify_and_get_state', process)
> state = process.request('get_state')
> assert state == modified_state
> assert state != initial_state
> ```

Could you add this example as test?

File: creepy/subprocess/pypen.py [hunk 5]
```diff
@@ -108,6 +131,26 @@ def detach(self):
...
+    if path is None:
+        path = os.path.join('/tmp', secrets.token_urlsafe(16))
+        os.mkfifo(path)
+    # Open the FIFO for reading. `os.O_NONBLOCK` is used in order to prevent blocking the process when opening.
+    in_fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
+    # Since we already opened the FIFO for reading, following open won't block.
+    out_fd = os.open(path, os.O_WRONLY)
+    # Set the file descriptor to just `O_RDONLY` in order to get rid of `O_NONBLOCK` and synchronize the execution of
+    # the two processes.
+    fcntl.fcntl(in_fd, fcntl.F_SETFL, os.O_RDONLY)
```
Comment (romawhite47) [inline creepy/subprocess/pypen.py:152]: Seems like the task to read and write to the same file should be know, is it standard solution?

Comment (veotani) [reply creepy/subprocess/pypen.py:160]: I haven't found anything. There are plenty examples of how to open pipes using `with open` syntax. However, it uses buffering and is not appropriate in our case.

**Pull request merged**